File: src/net/liplum/animations/anis/AniState.java
package net.liplum.animations.anis;

import arc.util.Nullable;
import mindustry.gen.Building;
import mindustry.world.Block;

/**
 * The animation state which decide how to render
 *
 * @param <TBlock> the type of block which has this animation state
 * @param <TBuild> the corresponding {@link Building} type
 */
public class AniState<TBlock extends Block, TBuild extends Building> {
    private final String stateName;
    @Nullable
    private final IRenderBehavior<TBuild> renderBehavior;
    private boolean overwriteBlock = false;

    /**
     * @param stateName a name
     */
    public AniState(String stateName) {
        this.stateName = stateName;
        this.renderBehavior = null;
    }

    /**
     * @param stateName a name
     * @param rb        how to render
     */
    public AniState(String stateName, IRenderBehavior<TBuild> rb) {
        this.stateName = stateName;
        this.renderBehavior = rb;
    }

    /**
     * Gets the name
     *
     * @return name
     */
    public String getStateName() {
        return stateName;
    }

    /**
     * Renders the current image
     *
     * @param build the subject to be rendered
     */
    public void drawBuilding(TBuild build) {
        if (this.renderBehavior != null) {
            this.renderBehavior.drawBuild(build);
        }
    }

    public boolean isOverwriteBlock() {
        return overwriteBlock;
    }

    public AniState<TBlock, TBuild> setOverwriteBlock(boolean overwriteBlock) {
        this.overwriteBlock = overwriteBlock;
        return this;
    }

    @Override
    public int hashCode() {
        return stateName.hashCode();
    }

    @Override
    public String toString() {
        return stateName;
    }
}

File: mdt/src/net/liplum/mdt/animations/anis/AniState.kt
package net.liplum.mdt.animations.anis

import arc.util.Nullable
import mindustry.gen.Building
import mindustry.world.Block
import java.util.function.Consumer

/**
 * The animation state which decide how to render
 *
 * @param <TBlock> the type of block which has this animation state
 * @param <TBuild> the corresponding [Building] type
</TBuild></TBlock> */
class AniState<TBlock : Block, TBuild : Building> {
    /**
     * Gets the name
     *
     * @return name
     */
    val stateName: String
    @Nullable
    var renderer: (TBuild.() -> Unit)?
        private set
    var isOverwriteBlock = false
        private set
    /**
     * @param stateName a name
     */
    constructor(stateName: String) {
        this.stateName = stateName
        renderer = null
    }
    /**
     * @param stateName a name
     * @param renderer  how to render
     */
    constructor(stateName: String, renderer: TBuild.() -> Unit) {
        this.stateName = stateName
        this.renderer = renderer
    }
    /**
     * @param stateName a name
     * @param renderer  how to render
     * For Java
     */
    constructor(stateName: String, renderer: Consumer<TBuild>) {
        this.stateName = stateName
        this.renderer = { renderer.accept(this) }
    }
    /**
     * Renders the current image
     *
     * @param build the subject to be rendered
     */
    fun drawBuilding(build: TBuild) {
        renderer?.invoke(build)
    }

    fun setOverwriteBlock(overwriteBlock: Boolean): AniState<TBlock, TBuild> {
        isOverwriteBlock = overwriteBlock
        return this
    }

    fun setRenderer(renderer: TBuild.() -> Unit): AniState<TBlock, TBuild> {
        this.renderer = renderer
        return this
    }

    override fun hashCode() =
        stateName.hashCode()

    override fun toString() =
        stateName

    override fun equals(other: Any?): Boolean {
        return when (other) {
            is AniState<*, *> -> other.stateName == this.stateName
            is String -> other == this.stateName
            else -> false
        }
    }
}